packages/compiler/src/server/tspconfig/completion.ts

```diff
@@ -211,3 +354,240 @@ export async function provideTspconfigCompletionItems(
+ * @param relativeFiles File relative path array
+ * @param siblings Sibling node array
+ * @returns CompletionItem object array
+ */
+function getFilePathCompletionItems(relativeFiles: string[], siblings: string[]): CompletionItem[] {
+  const items: CompletionItem[] = [];
+  if (relativeFiles.length > 0) {
+    for (const file of relativeFiles) {
+      if (!siblings.includes(file)) {
+        const item: CompletionItem = {
+          label: file,
+          kind: CompletionItemKind.File,
+        };
+        items.push(item);
+      }
+    }
+  }
+  return items;
+}
+
+/**
+ * resolve variable interpolation completion items
+ * @param target YamlScalarTarget object, contains the information of the target node
+ * @param tspConfigPosition current cursor position in the configuration file
+ * @param targetLine current line content of the target node
+ * @returns variable interpolation completion items
+ */
+function resolveVariableInterpolationCompleteItems(
+  target: YamlScalarTarget,
+  tspConfigPosition: Position,
+  targetLine: string,
+): CompletionItem[] {
+  const yamlDocNodes = target.yamlDoc;
+
+  if (target.sourceType === "QUOTE_DOUBLE" || target.sourceType === "QUOTE_SINGLE") {
+    const targetPos = targetLine.lastIndexOf("{");
+    const curText = targetLine.substring(targetPos, tspConfigPosition.character);
+
+    if (/{[^}]*env\.[^}]*$/.test(curText)) {
+      // environment-variables
+      return getCompletionItemsByFilter(
+        yamlDocNodes,
+        "environment-variables",
+        "Custom environment variables",
+      );
+    } else if (/{[^}]*$/.test(curText)) {
+      // parameters and built-in variables
+      const result = [
+        ...getCompletionItemsByFilter(yamlDocNodes, "parameters", "Custom paramters variables"),
+        ...["cwd", "project-root"].map((value) => {
+          const item: CompletionItem = {
+            label: value,
+            kind: CompletionItemKind.Value,
+            documentation: "Built-in variables",
+          };
+          return item;
+        }),
+      ];
+
+      // if the current path is options, add output-dir and emitter-name
+      if (target.path.length > 2 && target.path[0] === "options") {
+        result.push(
+          ...["output-dir", "emitter-name"].map((value) => {
+            const item: CompletionItem = {
+              label: value,
+              kind: CompletionItemKind.Value,
+              documentation: "Built-in variables",
+            };
+            return item;
+          }),
+        );
+      }
+      return result;
+    }
+  }
+
+  return [];
+}
+
+/**
+ * Get the corresponding CompletionItem array based on the filter name
+ * @param yamlDoc The contents of the YAML configuration file
+ * @param filterName The filter name，such as "parameters" or "environment-variables"
+ * @param description The description of the CompletionItem object
+ * @returns  CompletionItem object array
+ */
+function getCompletionItemsByFilter(
```

Comment:
this name is too big. It's better to change to something like getVariableCompletionItem() and change the filterName to varType: "parameter" | "environment-variable". and then you dont need the description because you have the enough knowledge to figure it out in the method.